test/index.js

```diff
@@ -1463,6 +1463,139 @@ describe('Nightmare', function () {
+        data.should.equal('mock-data');
+        done();
+      });
+    });
+
+    it('should support multiple concurrent frame subscriptions', function(done) {
+      var subscribeCount = 0;
+      var unsubscribeCount = 0;
+      var FrameManager = require('../lib/frame-manager.js');
+      var fn = null;
+      var assert = require('assert');
+      var async = require('async');
+      var manager = FrameManager({
+        webContents: {
+          debugger: {
+            isAttached: function() { return true; },
+            sendCommand: function(command) {
+              if (command === 'DOM.highlightRect') {
+                setTimeout(function () {
+                  fn('mock-data');
+                }, 100);
+              }
+            }
+          },
+          beginFrameSubscription: function(_fn) { subscribeCount += 1; assert.strictEqual(fn, null); fn = _fn; },
+          endFrameSubscription: function() { unsubscribeCount += 1; fn = null; },
+          executeJavaScript: function() {}
+        }
+      });
+      async.times(2, function requestFrameFn (i, cb) {
+        manager.requestFrame(function handleFrame (data) {
+          cb(null, data);
+        });
+      }, function handleResults (err, results) {
+        if (err) { done(err); }
+        subscribeCount.should.equal(1);
+        unsubscribeCount.should.equal(1);
+        results[0].should.equal('mock-data');
+        results[1].should.equal('mock-data');
+        done();
+      });
+    });
+
+    it('should support multiple series frame subscriptions', function(done) {
+      var subscribeCount = 0;
+      var unsubscribeCount = 0;
+      var FrameManager = require('../lib/frame-manager.js');
+      var fn = null;
+      var assert = require('assert');
+      var async = require('async');
+      var manager = FrameManager({
+        webContents: {
+          debugger: {
+            isAttached: function() { return true; },
+            sendCommand: function(command) {
+              if (command === 'DOM.highlightRect') {
+                setTimeout(function () {
+                  fn('mock-data');
+                }, 100);
+              }
+            }
+          },
+          beginFrameSubscription: function(_fn) { subscribeCount += 1; assert.strictEqual(fn, null); fn = _fn; },
+          endFrameSubscription: function() { unsubscribeCount += 1; fn = null; },
+          executeJavaScript: function() {}
+        }
+      });
+      async.timesSeries(2, function requestFrameFn (i, cb) {
+        manager.requestFrame(function handleFrame (data) {
+          cb(null, data);
+        });
+      }, function handleResults (err, results) {
+        if (err) { done(err); }
+        subscribeCount.should.equal(2);
+        unsubscribeCount.should.equal(2);
+        results[0].should.equal('mock-data');
+        results[1].should.equal('mock-data');
+        done();
+      });
+    });
+
+    // DEV: We can have multiple timeouts if page is static
```

Comment:
For what it’s worth, the entire point of #927 was to not have timeouts when the page is static ;)
(The timeout guard should really never fire at all, but is there just in case the debugger fails in ways it should probably not be able to or its API has a breaking change.)